framework/crud_resource.go

```diff
@@ -0,0 +1,403 @@
+	// informer knows about the old and the new custom object.
+	// GetDesiredState then receives the new custom object to be able to
+	// compute the desired state of a system.
+	GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error)
+
+	// NewUpdatePatch is called upon observed custom object change. It receives
+	// the observed custom object, the current state as provided by
+	// GetCurrentState and the desired state as provided by
+	// GetDesiredState. NewUpdatePatch analyses the current and desired
+	// state and returns the patch to be applied by Create, Delete, and
+	// Update functions. ApplyCreateChange, ApplyDeleteChange, and
+	// ApplyUpdateChange are called only when the corresponding patch part
+	// was created.
+	NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
+	// NewDeletePatch is called upon observed custom object deletion. It
+	// receives the deleted custom object, the current state as provided by
+	// GetCurrentState and the desired state as provided by
+	// GetDesiredState. NewDeletePatch analyses the current and desired
+	// state returns the patch to be applied by Create, Delete, and Update
+	// functions. ApplyCreateChange, ApplyDeleteChange, and
+	// ApplyUpdateChange are called only when the corresponding patch part
+	// was created.
+	NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
+
+	// ApplyCreateChange receives the new custom object observed during
+	// custom resource watches. It also receives the create portion of the
+	// Patch provided by NewUpdatePatch or NewDeletePatch.
+	// ApplyCreateChange only has to create resources based on its provided
+	// input. All other reconciliation logic and state transformation is
+	// already done at this point of the reconciliation loop.
+	ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error
+	// ApplyDeleteChange receives the new custom object observed during
+	// custom resource watches. It also receives the delete portion of the
+	// Patch provided by NewUpdatePatch or NewDeletePatch.
+	// ApplyDeleteChange only has to delete resources based on its provided
+	// input. All other reconciliation logic and state transformation is
+	// already done at this point of the reconciliation loop.
+	ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error
+	// ApplyUpdateChange receives the new custom object observed during
+	// custom resource watches. It also receives the update portion of the
+	// Patch provided by NewUpdatePatch or NewDeletePatch.
+	// ApplyUpdateChange has to update resources based on its provided
+	// input. All other reconciliation logic and state transformation is
+	// already done at this point of the reconciliation loop.
+	ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error
+}
+
+type CRUDResource struct {
+	ops CRUDResourceOps
+}
```

Comment:
This resource translates to the simpler one:

```
type Resource interface {
	EnsureCreated(ctx context.Context, obj interface{}) error
	EnsureDeleted(ctx context.Context, obj interface{}) error
}
```

So both of them can be used with the farmework.

Reply:
I'd like to replace current interface, and there is still work to do. I don't want to also change loggingresource and other stuff at once. Better to split that into PRs.